salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+    """
+    def __init__(self, opts, io_loop=None, timeout=60):
+        super(RMQNonBlockingConnectionWrapper, self).__init__(opts)
+
+        self._io_loop = tornado.ioloop.IOLoop.instance() if io_loop is None else io_loop
+        self._io_loop.make_current()
+
+        self._channel = None
+        self._timeout = timeout
+        self._callback = None
+
+        self._connection = SelectConnection(parameters=pika.ConnectionParameters(host=self._host, credentials=self._creds),
+                                                             on_open_callback=self._on_connection_open,
+                                                             on_open_error_callback=self._on_connection_error,
+                                                             custom_ioloop=self._io_loop)
+
+        # set up exchange for message broadcast
+        self._fanout_exchange_name = self.queue_name
+
+    def _on_connection_open(self, connection):
+        """
+        Invoked by pika when connection is opened successfully
+        :param connection:
+        :return:
+        """
+        self._channel = connection.channel(on_open_callback=self._on_channel_open)
+
+    def _on_connection_error(self, connection, exception):
+        """
+        Invoked by pika when connection on connection error
+        :param connection:
+        :param exception:
+        :return:
+        """
+        log.error("Failed to connect", exc_info=True)
+
+    def _on_channel_open(self, channel):
+        """
+        Invoked by pika when channel is opened successfully
+        :param channel:
+        :return:
+        """
+        self._channel.exchange_declare(exchange=self._fanout_exchange_name, exchange_type=ExchangeType.fanout, callback=self.on_exchange_declare)
+
+    def on_exchange_declare(self, method):
+        """Invoked by pika when RabbitMQ has finished the Exchange.Declare RPC
+        command.
+        """
+        log.debug("Exchange declared: {}".format(self._fanout_exchange_name))
+        self._channel.queue_declare(queue=self.queue_name, callback=self._on_queue_declared)
+
+    def _on_queue_declared(self, method):
+        """
+        Invoked by pika when queue is declared successfully
+        :param method:
+        :return:
+        """
+        log.debug("Queue declared: {}".format(self._queue_name))
+
+        self._channel.queue_bind(
+            self.queue_name,
+            self._fanout_exchange_name,
+            routing_key=self.queue_name,
+            callback=self._on_queue_bind)
+
+    def _on_queue_bind(self, method):
+        """
+        Invoked by pika when queue bound successfully
+        :param method:
+        :return:
+        """
+
+        log.debug("Queue bound [{}]".format(self.queue_name))
+
+        def _callback_wrapper(channel, method, properties, payload):
+             if self._callback:
+                log.debug("Calling message callback")
+                return self._callback(payload)
```

Comment:
**note** Not sure if it's desirable, but internal functions are (re)created each time the outer function is called, i.e. each call to `_on_queue_bind` will create a new copy of `_callback_wrapper` function. This is fine if it's only meant to be called once, but if the outer function is called 100x then you'll be creating 100x new instances of the function (that may or may not be discarded, depending on what references are kept :upside_down_face: )